cli/provision.py

```diff
@@ -17,71 +19,163 @@
+    for the application
+    """
 
-    # playbook execution stage
-    if not args['dry-run']:
-        vars(args)['settings'] = yaml.load(yaml.safe_dump(
-            cli.yamls.Lookup.settings, default_flow_style=False))
-        if not args['cleanup']:
-            vars(args)['provision'] = True
+    def __init__(self, mame, args, settings_dir):
+        self.name = mame
+        self.args = args
+        self.settings_dir = settings_dir
+
+    @classmethod
+    def create(cls, app, settings_dir, args):
+        """
+        Constructs the sub-command.
+        :param app:  tha application name
+        :param settings_dir:  the default application settings dir
+        :param args: the application args
+        :return: the IRSubCommand instance.
+        """
+        if args:
+            settings_dir = os.path.join(settings_dir,
+                                        app)
+        if hasattr(args, "command0"):
+            settings_dir = os.path.join(settings_dir,
+                                        args['command0'])
+        return cls(args['command0'], args, settings_dir)
+
+    def get_settings_files(self):
+        """
+        Collects all the settings files for given application and sub-command
+        """
+        settings_files = []
+
+        # first take all the files from the input-files args
+        for input_file in self.args['input-files'] or []:
+            settings_files.append(utils.normalize_file(input_file))
+
+        # get the sub-command yml file
+        settings_files.append(os.path.join(
+            self.settings_dir,
+            self.name + '.yml'))
+
+        # load directly from args
+        for key, val in vars(self.args).iteritems():
+            if val is not None and key not in NON_SETTINGS_OPTIONS:
+                settings_file = os.path.join(
+                    self.settings_dir, key, val + '.yml')
+                LOG.debug('Searching settings file for the "%s" key...' % key)
+                if not os.path.isfile(settings_file):
+                    settings_file = utils.normalize_file(val)
+                settings_files.append(settings_file)
+                LOG.debug('"%s" was added to settings '
+                          'files list as an argument '
+                          'for "%s" key' % (settings_file, key))
+
+        LOG.debug("All settings files to be loaded:\n%s" % settings_files)
+        return settings_files
+
+
+class IRApplication(object):
+    """
+    Hold the default application workflow logic.
+    """
+    def __init__(self, name, settings_dir, args):
+        self.name = name
+        self.args = args.args
+        self.settings_dir = settings_dir
+        self.sub_command = IRSubCommand.create(name, settings_dir, args)
+
+    def run(self):
+        """
+        :return: Runs the application
+        """
+        settings = self.lookup(self.sub_command.get_settings_files())
+        self.dump_settings(settings)
+        self.execute(settings)
```

Comment:
This is not a blocker, but each application should determine its own execute method and subcommands might wish to call a different playbook than default.